### ISSUE TITLE
NixOS evaluation speed regression

### ISSUE THREAD

**Issue by edolstra** [NixOS/nixpkgs#79943]
**Describe the bug**

NixOS evaluation has gotten ~3x slower since 18.09.

**To Reproduce**

    $ command time nix-instantiate nixos/release-combined.nix -A nixos.tests.misc.x86_64-linux --dry-run

on my system (Ryzen 3700X) takes ~1.7s on 18.09, but ~5.5s on master. And for historical perspective, it takes 0.4s on 14.04.

Also, memory usage went from 113 MiB max resident in 14.04, to 395 MiB in 18.09, and 666 MiB on master. (This is with GC_INITIAL_HEAP_SIZE=1m.)

**Comment by edolstra** [comment]
Maybe it's time to revisit https://github.com/NixOS/rfcs/pull/22.

**Comment by grahamc** [comment]
I'm hopeful that Flakes will make 22/(something like 22) an easy, obvious choice for growth for 20.09.

**Comment by edolstra** [comment]
Flakes help with out-of-tree modules, but they're not necessary for modules that are in Nixpkgs but not in `module-list.nix`. Those can be imported using `imports = [ (pkgs.path + "/nixos/modules/.../bla.nix") ]`.

**Comment by grahamc** [comment]
Yeah,  think we could create a way for people to contribute modules somewhere organized by the community and fetched via flakes -- instead of inclusion in to NixOS. One subtlety of my previous comment I want to point out is that I don't think we can make such changes for 20.03.

**Comment by Infinisil** [comment]
Related is https://github.com/NixOS/nixpkgs/issues/57477

**Comment by roberth** [comment]
I've written down my thoughts in [NixOS always imports all modules]( https://github.com/NixOS/nixpkgs/issues/137168#issuecomment-917257915).

It's a low tech solution, which makes it viable. Also with flakes we'll need the same transformation to happen anyway.

**Comment by infinisil** [comment]
Still a problem
